eng/scripts/distribute-packages-to-matrix.ps1

```diff
@@ -0,0 +1,263 @@
+    Write-Error "PackageInfo folder file not found: $PackageInfoFolder"
+    exit 1
+}
+
+if (!(Test-Path $PlatformMatrix)) {
+    Write-Error "Platform matrix file not found: $PlatformMatrix"
+    exit 1
+}
+
+function Extract-MatrixMultiplier {
+    param (
+        [Parameter(Mandatory=$true)]
+        [PSCustomObject]$Matrix
+    )
+
+    $highestCount = 1
+
+    foreach ($property in $Matrix.PSObject.Properties) {
+        $type = $property.Value.GetType().Name
+        switch ($type) {
+            "PSCustomObject" {
+                $itemCount = 0
+
+                # this looks very strange to loop over the properties of a PSCustomObject,
+                # but this is the only way to actually count. The Count/Length property is
+                # NOT available
+                foreach($innerProperty in $property.Value.PSObject.Properties) {
+                    $itemCount++
+                }
+
+                if ($itemCount -gt $highestCount) {
+                    $highestCount = $itemCount
+                }
+            }
+            "Object[]" {
+                $count = $property.Value.Length
+                if ($count -gt $highestCount) {
+                    $highestCount = $count
+                }
+            }
+        }
+    }
+
+    return $highestCount
+}
+
+function Split-ArrayIntoBatches {
+    param (
+        [Parameter(Mandatory=$true)]
+        [object[]]$InputArray,
+
+        [Parameter(Mandatory=$true)]
+        [int]$BatchSize
+    )
+
+    $batches = @()
+
+    for ($i = 0; $i -lt $InputArray.Count; $i += $BatchSize) {
+        $batch = $InputArray[$i..[math]::Min($i + $BatchSize - 1, $InputArray.Count - 1)]
+
+        $batches += ,$batch
+    }
+
+    return ,$batches
+}
+
+function Update-Include {
+    param (
+        [Parameter(Mandatory=$true)]
+        $Matrix,
+        [Parameter(Mandatory=$true)]
+        $IncludeConfig,
+        [Parameter(Mandatory=$true)]
+        $TargetingString
+    )
+
+    foreach ($configElement in $IncludeConfig) {
+        if ($configElement.PSObject.Properties) {
+            $topLevelPropertyName = $configElement.PSObject.Properties.Name
+            $topLevelPropertyValue = $configElement.PSObject.Properties[$topLevelPropertyName].Value
+
+            if ($topLevelPropertyValue.PSObject.Properties) {
+                $secondLevelPropertyName = $topLevelPropertyValue.PSObject.Properties.Name
+                $secondLevelPropertyValue = $topLevelPropertyValue.PSObject.Properties[$secondLevelPropertyName].Value
+
+                $newSecondLevelName = "$secondLevelPropertyName" + "$TargetingString"
+
+                $topLevelPropertyValue.PSObject.Properties.Remove($secondLevelPropertyName)
+                $topLevelPropertyValue | Add-Member -MemberType NoteProperty -Name $newSecondLevelName -Value $secondLevelPropertyValue
+
+                if (-not $topLevelPropertyValue.$newSecondLevelName.PSObject.Properties["TargetingString"]) {
```

Comment:
follow-up: As we start to work on the other languages I think this matrix script will need some more generalizations so that we can hopefully handle this expansion in more shared manor.